### PR TITLE
Suggest Explicit Lifetime for Associated Type Bindings

### LINKED ISSUE #122025: Missing explicit lifetime in associated type suggests introducing invalid higher-ranked lifetime
### Code

```Rust
trait T {
    type Assoc;
}

fn foo<F>(t: F) where F: T<Assoc=&str> {}
```

### Current output

```Shell
error[E0637]: `&` without an explicit lifetime name cannot be used here
 --> src/main.rs:5:34
  |
5 | fn foo<F>(t: F) where F: T<Assoc=&str> {}
  |                                  ^ explicit lifetime name needed here
  |
help: consider introducing a higher-ranked lifetime here
  |
5 | fn foo<F>(t: F) where F: for<'a> T<Assoc=&'a str> {}
  |                          +++++++          ++
```

### Desired output

```Shell
error[E0637]: `&` without an explicit lifetime name cannot be used here
 --> src/main.rs:5:34
  |
5 | fn foo<F>(t: F) where F: T<Assoc=&str> {}
  |                                  ^ explicit lifetime name needed here
  |
help: consider adding an explicit lifetime here
  |
5 | fn foo<'a, F>(t: F) where F: T<Assoc=&'a str> {}
  |        ++                             ++
```

### Rationale and extra context

If you try to follow the suggested help text, it just gets you another error:
```
error[E0582]: binding for associated type `Assoc` references lifetime `'a`, which does not appear in the trait input types
  --> src/lib.rs:17:36
   |
5  | fn foo<F>(t: F) where F: for<'a> T<Assoc=&'a str> {}
   |                                    ^^^^^^^^^^^^^
```
i.e. it's not the correct course of action here

### Other cases

_No response_

### Rust Version

```Shell
❯ rustc --version --verbose
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-pc-windows-msvc
release: 1.76.0
LLVM version: 17.0.6

But also verified to behave the same on nightly (2024-03-04 d18480b84fdbf1efc34f) on the playground
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by veera-sivarajan**
Fixes #122025 

This PR suggests an explicit lifetime annotation for associated type bindings. Previously, this error suggested an incorrect HRTB.

Please let me know if there's a better way to figure `is_associated_type_binding`. Also, this suggestion uses `Applicability::MaybeIncorrect` as the suggestion can be incorrect when a trait's input lifetimes are also missing.

### CODE REVIEW

File: compiler/rustc_resolve/src/late.rs [hunk 1]
```diff
@@ -1685,14 +1684,53 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
...
+                                } else {
+                                    diag.multipart_suggestion(
+                                        "consider introducing a higher-ranked lifetime here",
+                                        vec![
+                                            (span.shrink_to_lo(), "for<'a> ".into()),
+                                            (lifetime.ident.span.shrink_to_hi(), "'a ".into()),
+                                        ],
+                                        Applicability::MachineApplicable,
+                                    );
+                                }
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late.rs:1733]: There is existing diagnostic code in `rustc_resolve::late::diagnostics`. We should adapt and reuse `add_missing_lifetime_specifiers_label` and `suggest_introducing_lifetime`.

**Comment by fmease** [comment]
CC https://github.com/rust-lang/rust/issues/123713#issuecomment-2047308830. The “proper” way to fix #122025 is to *stash* the diagnostic in `rustc_resolve::late` and `try_steal_modify_and_emit_err` it during HIR ty lowering where we have the necessary tools available to check if a lifetime param would be unconstrained. However that might not be feasible, needs experimentation.

I've only *glanced* at the changes made in this PR but you seem to be taking an alternative approach I've dubbed “heuristc in `late`” in my linked comment. That's perfectly valid, too. My linked comments might provide some more food for thought :)

**Comment by Dylan-DPC** [comment]
@veera-sivarajan any updates on this? thanks

**Comment by veera-sivarajan** [comment]
Still working on it. Thanks for checking.

**Pull request closed without merging**